demos/common/processing/async_pipeline.h

```diff
@@ -0,0 +1,150 @@
+/// Derived classes should add functions for data submission and output processing
+class PipelineBase
+{
+public:
+    struct PerformanceInfo
+    {
+        int64_t framesCount = 0;
+        std::chrono::steady_clock::duration latencySum = std::chrono::steady_clock::duration::zero();
+        std::chrono::steady_clock::duration lastInferenceLatency = std::chrono::steady_clock::duration::zero();
+        std::chrono::steady_clock::time_point startTime;
+        double movingAverageLatencyMs;
+        uint32_t numRequestsInUse;
+        double movingAverageFPS;
+        double FPS=0;
+
+        double getTotalAverageLatencyMs() const {
+            return ((double)std::chrono::duration_cast<std::chrono::milliseconds>(latencySum).count()) / framesCount;
+        }
+
+        double getLastInferenceLatencyMs() const {
+            return ((double)std::chrono::duration_cast<std::chrono::milliseconds>(lastInferenceLatency).count());
+        }
+
+    };
+
+    struct PerformanceInternalCounters
+    {
+        long long latenciesMs[MOVING_AVERAGE_SAMPLES] = {};
+        std::chrono::steady_clock::time_point retrievalTimestamps[MOVING_AVERAGE_SAMPLES] = {};
+        long long movingLatenciesSumMs = 0;
+        int currentIndex = 0;
+    };
+
+public:
+    /// Loads model and performs required initialization
+    /// @param modelInstance pointer to model object. Object it points to should not be destroyed manually after passing pointer to this function.
+    /// @param cnnConfig - fine tuning configuration for CNN model
+    /// @param engine - pointer to InferenceEngine::Core instance to use.
+    /// If it is omitted, new instance of InferenceEngine::Core will be created inside.
+    PipelineBase(ModelBase* modelInstance, const CnnConfig& cnnConfig, InferenceEngine::Core* engine = nullptr);
+    virtual ~PipelineBase();
+
+    /// Waits until either output data becomes available or pipeline allows to submit more input data.
+    /// @param shouldKeepOrder if true, function will treat results as ready only if next sequential result (frame) is
+    /// ready (so results can be extracted in the same order as they were submitted). Otherwise, function will return if any result is ready.
+    void waitForData(bool shouldKeepOrder = true);
+
+    /// Returns true if there's available infer requests in the pool
+    /// and next frame can be submitted for processing.
+    /// @returns true if there's available infer requests in the pool
+    /// and next frame can be submitted for processing, false otherwise.
+    bool isReadyToProcess() { return requestsPool->isIdleRequestAvailable(); }
+
+    /// Returns performance info
+    /// @returns performance information structure
+    PerformanceInfo getPerformanceInfo() { std::lock_guard<std::mutex> lock(mtx); return perfInfo; }
+
+    /// Waits for all currently submitted requests to be completed.
+    ///
+    void waitForTotalCompletion() { if(requestsPool.get())requestsPool->waitForTotalCompletion(); }
```

Comment:
I don't see why the function is needed. Just wait in `requestsPool`'s destructor.

Reply:
ok. fixed

Reply:
Reorder fields.

Reply:
See comments below

Reply:
as we agreed, we leave this as is.